tests/datavolume_test.go

```diff
@@ -471,4 +474,147 @@ var _ = Describe("DataVolume Integration", func() {
+					err := virtClient.VirtualMachine(createdVirtualMachine.Namespace).Delete(createdVirtualMachine.Name, &metav1.DeleteOptions{})
+					Expect(err).ToNot(HaveOccurred())
+				}
+
+				if dataVolume != nil {
+					err := virtClient.CdiClient().CdiV1alpha1().DataVolumes(dataVolume.Namespace).Delete(dataVolume.Name, &metav1.DeleteOptions{})
+					Expect(err).ToNot(HaveOccurred())
+				}
+			})
+
+			It("deny then allow clone request", func() {
+				vm := tests.NewRandomVMWithCloneDataVolume(dataVolume.Namespace, dataVolume.Name, tests.NamespaceTestDefault)
+
+				vmBytes, err := json.Marshal(vm)
+				Expect(err).ToNot(HaveOccurred())
+				byteReader := bytes.NewReader(vmBytes)
+
+				// this should fail because don't have permission
+				stdOut, stdErr, err := tests.RunCommandWithNSAndInput(vm.Namespace, byteReader, "kubectl", "create", "--as", serviceAccount, "-f", "-")
+				if err == nil {
+					fmt.Printf("command should have failed\nstdOut\n%s\nstdErr\n%s\n", stdOut, stdErr)
+					Expect(err).To(HaveOccurred())
+				}
+				Expect(stdErr).Should(ContainSubstring("Authorization failed, message is:"))
+
+				// add permission
+				cloneRole, cloneRoleBinding = addClonePermission(virtClient, tests.AdminServiceAccountName, tests.NamespaceTestDefault, tests.NamespaceTestAlternative)
+
+				// sometimes it takes a bit for permission to actually be applied so eventually
+				Eventually(func() bool {
+					byteReader = bytes.NewReader(vmBytes)
+					stdOut, stdErr, err = tests.RunCommandWithNSAndInput(vm.Namespace, byteReader, "kubectl", "create", "--as", serviceAccount, "-f", "-")
+					if err != nil {
+						fmt.Printf("command should have succeeded maybe new permissions not applied yet\nstdOut\n%s\nstdErr\n%s\n", stdOut, stdErr)
+						return false
+					}
+					return true
+				}, 90*time.Second, 1*time.Second).Should(BeTrue())
+
+				createdVirtualMachine = vm
+
+				// wait for clone to complete
+				targetDVName := vm.Spec.DataVolumeTemplates[0].Name
+				Eventually(func() bool {
+					dv, err := virtClient.CdiClient().CdiV1alpha1().DataVolumes(createdVirtualMachine.Namespace).Get(targetDVName, metav1.GetOptions{})
+					if err != nil && errors.IsNotFound(err) {
+						return false
+					}
+					Expect(err).ToNot(HaveOccurred())
+					return dv.Status.Phase == cdiv1.Succeeded
+				}, 90*time.Second, 1*time.Second).Should(BeTrue())
+
+				// start/stop vm
+				createdVirtualMachine = tests.StartVirtualMachine(createdVirtualMachine)
+				createdVirtualMachine = tests.StopVirtualMachine(createdVirtualMachine)
+			})
+		})
+	})
 })
+
+func addClonePermission(client kubecli.KubevirtClient, sa, saNamespace, targetNamesace string) (*rbacv1.Role, *rbacv1.RoleBinding) {
+	var err error
+	resourceName := sa + "-cloner"
+
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: resourceName,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{
+					"cdi.kubevirt.io",
+				},
+				Resources: []string{
+					"datavolumes/source",
```

Comment:
I'm not sure I understand this rule. Why isn't the permission being checked/validated the ability to access PVCs on the other namespace rather than the datavolumes/source resource?

Reply:
Yeah, this is something I was going back and forth on.  I think the closest thing to "access PVC" permission is "create pod".  However, the clone pods are currently privileged.  And I couldn't find out how to determine if a user can create privileged pods with upstream k8s.  Three is an api on OpenShift, however.  So I figured it would be best to do something that kind of satisfies both cases.  But maybe something else would be better?  Totally open to reevaluating this.  Especially when network cloning comes around.

You also bring up a good point regarding serviceaccounts.  I blindly just implemented what we do in CDI, and since datavolumes don't have service accounts, we have to check user permissions in a webhook.  If we decide to go with serviceaccounts here, would we still do the authorization check in a webhook?  Or should the controller handle it?

Reply:
> However, the clone pods are currently privileged. And I couldn't find out how to determine if a user can create privileged pods with upstream k8s. 

Yeah, that's what Pod Security Policies (PSPs) introduce. It's the equivalent of OpenShift's SCC. I'm unsure if that's relevant here or not though. 

The fact that the the clone pod is privileged probably doesn't matter in this context. The user doesn't have to be able to create a privileged pod in another namespace in order to have access to the PVC in another namespace.

Essentially what we want to know is if the user/SA has the permissions to launch a pod that can access the PVC's contents in question. If they can do that, then we know they're allowed to view the PVC's contents... and if they can view the contents, then we should be able to draw the conclusion that they are allowed to do whatever they want with those contents, including copy the data to another PVC in another namespace.


> You also bring up a good point regarding serviceaccounts. I blindly just implemented what we do in CDI, and since datavolumes don't have service accounts, we have to check user permissions in a webhook. If we decide to go with serviceaccounts here, would we still do the authorization check in a webhook? Or should the controller handle it?

The way you have it now is fine. You're checking to see if the serviceaccount (same thing as the user) provided in the Create request that the webhook intercepts is allowed to create the DataVolume that does cross namespace cloning.

Reply:
> Yeah, that's what Pod Security Policies (PSPs) introduce. It's the equivalent of OpenShift's SCC. I'm unsure if that's relevant here or not though.

I understand that upstream has PSPs but they do not have this nice api that openshift has that can tell you whether a user can create a privileged pod:  https://github.com/openshift/api/blob/master/security/v1/types.go#L297-L336

>The fact that the the clone pod is privileged probably doesn't matter in this context. The user doesn't have to be able to create a privileged pod in another namespace in order to have access to the PVC in another namespace. 

Do PSPs keep a user from execing into a privileged pod if they are not authorized to create privileged pods?  As far as I know, they do not.  But I know this is not an issue on OpenShift with SCCs.  So, my concern is that an upstream user could start a clone and exec into a privileged pod.  

>Essentially what we want to know is if the user/SA has the permissions to launch a pod that can access the PVC's contents in question. If they can do that, then we know they're allowed to view the PVC's contents... and if they can view the contents, then we should be able to draw the conclusion that they are allowed to do whatever they want with those contents, including copy the data to another PVC in another namespace.

I agree, but the side effect of the pod being privileged makes things a little more complicated IMO.  Which is how we ended up with the "datavolume/source" resource.  Presumably, an admin would have to explicitly grant this permission, hopefully with the knowledge that it could lead to privilege escalation for the user.

This will all be moot pretty soon once network cloning is done and the pods are no longer privileged.  Then the webhook can check "create pod".  And I guess we'll leave "datavolume/source" around for awhile too.

Reply:
> Do PSPs keep a user from execing into a privileged pod if they are not authorized to create privileged pods

no, but SCCs appear to when I tested this a few weeks ago. 

> So, my concern is that an upstream user could start a clone and exec into a privileged pod

if this isn't on OpenShift, then that would be the case. 

> This will all be moot pretty soon once network cloning is done and the pods are no longer privileged. Then the webhook can check "create pod". And I guess we'll leave "datavolume/source" around for awhile too

Is it worth adding any special consideration beyond "pod create" privileges here if the network change is landing soon? The primary concern seems to be around pod/exec into a privileged pod. SCCs prevent that in OpenShift, and the networking change fixes it for everything once it lands in CDI. 

also, unless there's a validation webhook checking PVC creation, is it not possible for a user to create a PVC with specific annotations that result in privileged pods being created for the clone as well?

Reply:
I think "pod create" will be sufficient when network cloning lands.

There is no validation webhook on PVC creation (didn't want to be a bottleneck/potential source of instability to a core resource) but the datavolume webhook adds a jwt token similar to the upload token as an annotation to the pvc.  The clone controller wont do anything if the annotation doesn't exist or is invalid.

Reply:
> The way you have it now is fine. You're checking to see if the serviceaccount (same thing as the user) provided in the Create request that the webhook intercepts is allowed to create the DataVolume that does cross namespace cloning.

What interests me here is, if we would create a controller which manages VMs, would we in the current implementation use the permissions of the controller, the serviceaccount on the VM, or of the user who created a CR for the VM-managing-controller?

Reply:
In the current implementation, the KubeVirt webhook checks that the user creating the VM CR has permission to clone.  We could alternatively check the service account of the VM.  Or both.

When the DataVolume is actually created by the controller in KubeVirt, a webhook in CDI will check that the controller has permission to clone.  Which should of course be true.  Which is why we need the first check.

Reply:
> When the DataVolume is actually created by the controller in KubeVirt, a webhook in CDI will check that the controller has permission to clone. Which should of course be true. Which is why we need the first check.

That is the issue I see. I guess we would have to check the service account, because even if we would revoke permissions for the service account or the user, VMI restarts would still happily clone from there. If the clone-source got exchanged, that means that you can access and extract data from a place where you should not have access to.

Reply:
I see.  I did not know that the clone operation may be executed multiple times on the kubevirt side.  The KubeVirt controller will have to do clone permission checking before creating the datavolume.  This makes the webhook check unnecessary.  Although, I think it does make for a better user experience.  Should we keep it?

Reply:
> I see. I did not know that the clone operation may be executed multiple times on the kubevirt side.

Yes, just delete the PVC and the DataVolume and it will be recreated by the controller, which involves a new clone.

> The KubeVirt controller will have to do clone permission checking before creating the datavolume.

This would bean that we go with the service account on the VMI in this scenario, right? Since there is nothing like a mechanism which tracks who created an object ...

Reply:
Yeah would have to check the service account